File: examples/solve-cloudflare-challenges/frameworks/playwright/java/CloudflareUnblock.java
// Bypasses Cloudflare challenges via /unblock, then connects Playwright to the
// already-unblocked session using the returned browserWSEndpoint.
// The ttl parameter keeps the browser alive long enough for the client to connect.
//
// Add to pom.xml:
//   <dependency>
//     <groupId>com.microsoft.playwright</groupId>
//     <artifactId>playwright</artifactId>
//     <version>1.44.0</version>
//   </dependency>
//   <dependency>
//     <groupId>com.google.code.gson</groupId>
//     <artifactId>gson</artifactId>
//     <version>2.10.1</version>
//   </dependency>
//
// Compile: mvn compile
// Run:     mvn exec:java

import com.google.gson.JsonParser;
import com.microsoft.playwright.*;
import java.net.URI;
import java.net.http.*;

public class CloudflareUnblock {
    public static void main(String[] args) throws Exception {
        String TOKEN = "YOUR_API_TOKEN_HERE";
        HttpClient client = HttpClient.newHttpClient();

        String body = "{\"url\":\"https://example-cloudflare-protected.com\","
            + "\"browserWSEndpoint\":true,\"ttl\":30000}";
        HttpRequest req = HttpRequest.newBuilder()
            .uri(URI.create("https://production-sfo.browserless.io/unblock?token=" + TOKEN + "&proxy=residential"))
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(body))
            .build();
        HttpResponse<String> httpRes = client.send(req, HttpResponse.BodyHandlers.ofString());

        // The /unblock response returns a raw WebSocket URL — append the token before connecting.
        String wsEndpoint = JsonParser.parseString(httpRes.body()).getAsJsonObject()
            .get("browserWSEndpoint").getAsString() + "?token=" + TOKEN;

        try (Playwright playwright = Playwright.create()) {
            Browser browser = playwright.chromium().connectOverCDP(wsEndpoint);
            try {
                BrowserContext context = browser.contexts().get(0);
                Page page = context.pages().get(0);
                System.out.println("Title: " + page.title());
                System.out.println("URL: " + page.url());
            } finally {
                // Always close to release the session even on error.
                browser.close();
            }
        }
    }
}

File: examples/solve-cloudflare-challenges/frameworks/playwright/csharp/CloudflareUnblock.cs
// Bypasses Cloudflare challenges via /unblock, then connects Playwright to the
// already-unblocked session using the returned browserWSEndpoint.
// The ttl parameter keeps the browser alive long enough for the client to connect.
//
// Install: dotnet add package Microsoft.Playwright
// Run:     dotnet run

using Microsoft.Playwright;
using System;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;

class CloudflareUnblock
{
    static async Task Main()
    {
        const string TOKEN = "YOUR_API_TOKEN_HERE";
        using var httpClient = new HttpClient();

        var body = new StringContent(
            "{\"url\":\"https://example-cloudflare-protected.com\",\"browserWSEndpoint\":true,\"ttl\":30000}",
            Encoding.UTF8, "application/json");
        var res = await httpClient.PostAsync(
            $"https://production-sfo.browserless.io/unblock?token={TOKEN}&proxy=residential",
            body);

        using var json = JsonDocument.Parse(await res.Content.ReadAsStringAsync());
        // The /unblock response returns a raw WebSocket URL — append the token before connecting.
        string wsEndpoint = json.RootElement.GetProperty("browserWSEndpoint").GetString()
            + $"?token={TOKEN}";

        using var playwright = await Playwright.CreateAsync();
        var browser = await playwright.Chromium.ConnectOverCDPAsync(wsEndpoint);
        try
        {
            var context = browser.Contexts[0];
            var page = context.Pages[0];
            Console.WriteLine($"Title: {await page.TitleAsync()}");
            Console.WriteLine($"URL: {page.Url}");
        }
        finally
        {
            // Always close to release the session even on error.
            await browser.CloseAsync();
        }
    }
}
